File: data/sliding-window/binary-subarrays-sum.cpp
#include <vector>

using namespace std;

int atMost(vector<int>& nums, int k) {
    if (k < 0) return 0;
    int left = 0, currentSum = 0, count = 0;
    for (int right = 0; right < nums.size(); ++right) {
        currentSum += nums[right];
        while (currentSum > k) {
            currentSum -= nums[left++];
        }
        count += right - left + 1;
    }
    return count;
}

int numSubarraysWithSum(vector<int>& nums, int goal) {
    return atMost(nums, goal) - atMost(nums, goal - 1);
}

File: data/sliding-window/binary-subarrays-sum.py
def numSubarraysWithSum(nums: list[int], goal: int) -> int:
    def atMost(k):
        if k < 0: return 0
        left = 0
        current_sum = 0
        count = 0
        for right in range(len(nums)):
            current_sum += nums[right]
            while current_sum > k:
                current_sum -= nums[left]
                left += 1
            count += right - left + 1
        return count

    return atMost(goal) - atMost(goal - 1)
